mne_sandbox/preprocessing/_sns.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+
+import numpy as np
+
+from mne import compute_raw_covariance
+from mne.io.pick import _pick_data_channels, pick_channels
+from mne.io import _BaseRaw
+from mne.utils import logger, verbose
+
+from ._dss import _pca
+
+
+class SensorNoiseSuppression(object):
+    """Apply the sensor noise suppression (SNS) algorithm
+
+    This algorithm (from [1]_) will replace the data from each channel by
+    its regression on the subspace formed by the other channels.
+
+    .. note:: Bad channels are not modified or reset by this function.
+
+    Parameters
+    ----------
+    raw : Instance of Raw
+        The raw data to process.
+    n_neighbors : int
+        Number of neighbors (based on correlation) to include in the
+        projection.
+    reject : dict | str | None
+        Rejection parameters based on peak-to-peak amplitude.
+        Valid keys are 'grad' | 'mag' | 'eeg' | 'eog' | 'ecg'.
+        If reject is None then no rejection is done.
+        This is only used during the covariance-fitting phase.
+    flat : dict | None
+        Rejection parameters based on flatness of signal.
+        Valid keys are 'grad' | 'mag' | 'eeg' | 'eog' | 'ecg', and values
+        are floats that set the minimum acceptable peak-to-peak amplitude.
+        If flat is None then no rejection is done.
+        This is only used during the covariance-fitting phase.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    raw : instance of Raw
+        The cleaned raw data.
+    operator : ndarray, shape (n_meg_ch, n_meg_ch)
+        The spatial operator that was applied to the MEG channels.
+
+    References
+    ----------
+    .. [1] De Cheveigné A, Simon JZ. Sensor noise suppression. Journal of
+           Neuroscience Methods 168: 195–202, 2008.
+    """
+
+    @verbose
+    def __init__(self, raw, n_neighbors, reject=None, flat=None, verbose=None):
```

Comment:
@agramfort confirms that it should be `sns.fit(raw)` and the rest should remain in `__init__()`

also most of the logic should go to `fit()` I think.